### ISSUE TITLE
[QST] Any practical upper limit on the number of items

### ISSUE THREAD

**Issue by lightsailpro** [NVIDIA-Merlin/Transformers4Rec#427]
# ❓ Questions & Help

## Details
Is there any limit on the number of items supported? For example, will the model cause any issue if the number of items is in the million scale? Thanks

**Comment by gabrielspmoreira** [comment]
Hi @lightsailpro. Yes, currently Transformers4Rec does not support model parallel training, which means that the model size is limited to single GPU memory. I did some experiments in the past with a V100 GPU 32GB and the maximum cardinality for the item id (with embedding size of 512) was in the order of 9-10M before getting Out-Of-Memory. You can use A100 GPUs which has 40GB or 80GB to store larger embedding tables. In addition to that limitation, Transformers4Rec does not use negative sampling to compute the loss function, it scores all items at each training/prediction step.
So definitely Transformers4Rec is currently not suitable for large scale item catalogs (tens of million items).
If you have such scenario I recommend you checking [HugeCTR](https://github.com/NVIDIA-Merlin/HugeCTR), which shards embeddings tables into multiple GPU and CPU Hosts and have smart strategies for model parallel training.
Futhermore, on next releases of [Merlin Models](https://github.com/NVIDIA-Merlin/models) (for TF API) we are going to integrate support for Sequential recommendation (including Transformers) and also distributed embedding tables

**Comment by lightsailpro** [comment]
Thanks @gabrielspmoreira. That helps a lot. Sequential recommendation support will be great.

**Comment by silpara** [comment]
@gabrielspmoreira what were the memory optimisations/parameter settings needed to scale to 9-10M items with 512 dimension embeddings? E.g. did you had to choose a smaller sequence length, batch size, or optimise how memory was allocated to NvTabular vs PyTorch etc?
